controllers/reconcilers/kserve_networkpolicy_reconciler.go

```diff
@@ -98,47 +104,128 @@ func (r *KserveNetworkPolicyReconciler) createDesiredResource(isvc *kservev1beta
 			PolicyTypes: []v1.PolicyType{v1.PolicyTypeIngress},
 		},
 	}
-	return desiredNetworkPolicy, nil
+	return desiredNetworkPolicy
 }
 
-func (r *KserveNetworkPolicyReconciler) getExistingResource(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) (*v1.NetworkPolicy, error) {
-	return r.networkPolicyHandler.FetchNetworkPolicy(ctx, log, types.NamespacedName{Name: networkPolicyName, Namespace: isvc.Namespace})
+// allowOpenshiftIngressPolicy creates policy that grants ingress access through Openshift Routes to the services which
+// are part of a namespace under Service Mesh, but are not under Service Mesh control.
+func (r *KserveNetworkPolicyReconciler) allowOpenshiftIngressPolicy(isvc *kservev1beta1.InferenceService) *v1.NetworkPolicy {
+	return &v1.NetworkPolicy{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "NetworkPolicy",
+			APIVersion: "networking.k8s.io/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      openshiftIngressNetworkPolicyName,
+			Namespace: isvc.Namespace,
+			Labels:    map[string]string{"opendatahub.io/related-to": "RHOAIENG-1003"},
+		},
+		Spec: v1.NetworkPolicySpec{
+			PodSelector: metav1.LabelSelector{},
+			Ingress: []v1.NetworkPolicyIngressRule{
+				{
+					From: []v1.NetworkPolicyPeer{
+						{
+							NamespaceSelector: &metav1.LabelSelector{
+								MatchLabels: map[string]string{"network.openshift.io/policy-group": "ingress"},
+							},
+						},
+					},
+				},
+			},
+			PolicyTypes: []v1.PolicyType{v1.PolicyTypeIngress},
+		},
+	}
 }
 
-func (r *KserveNetworkPolicyReconciler) processDelta(ctx context.Context, log logr.Logger, desiredPod *v1.NetworkPolicy, existingPod *v1.NetworkPolicy) (err error) {
+// allowTrafficFromApplicationNamespaces creates combined network policy applied to pods in InferenceService namespace.
+// This set of policies allow traffic from:
+// - application namespace, where OpenDataHub and component services are deployed
+// - namespaces created by OpenDataHub where components live
+// - traffic from other DataScienceProjects (namespaces created through dashboard)
```

Comment:
User can create workloads in other OCP projects which were not originally created from Dashboard.  it's a recent feature.

I can't tell how much it is used, but still, it's a scenario. Would this be a problem? 

![image](https://github.com/opendatahub-io/odh-model-controller/assets/88311595/6b3b0e18-f8d3-4e33-8c6d-8c780c827db8)

Reply:
@bdattoma Then it boils down to the question if it is possible to know that namespace X from where the traffic is originating is part of the Open Data Hub. Is there any label that would help us answer this question? If not, then it would be blocked even with 2.5 and KServe only, due to e.g. already existing `allow-openshift-monitoring-ns` policy.

Reply:
Can anyone shed some light here? @danielezonca @zdtsw

Reply:
I think, but should be confirmed, that pods created via dashboard have the `"opendatahub.io/dashboard": "true"` label. I've checked with a workbench created from UI but running into a OCP project (not created via RHOAI Dashboard).

So, could we use a such pod selector in the NP?

```
podSelector:
        matchLabels:
          "opendatahub.io/dashboard": "true"
```

Reply:
hm I see now that Modelmesh model pods, for example, don't have the same label ^. It has something else like `modelmesh-service=modelmesh-serving`

Reply:
i think the best thing is to filter on which namespace, and have the open for all pods running in these namespaces first, then we can set more strict match on the pod level or even ports etc. 

the question in this comment is, if user can select "All available projects" from the UI dropdown, then it is much more namespace can be used as a datascience project(?) (in my dev cluster, i can even select istio-system and knative-eventing,....)
P.S:  i did not know this is available till now :woman_shrugging:

Reply:
> then it is much more namespace can be used as a datascience project(?)


correct.

If having a generic solution here is complicated and not wroth the effort, we could probably make it working for the proper DS Projects (i.e., from UI), and release note a workaround in the case the user is using a generic OCP project. I don't link much as approach, but:
- I'm unsure about likelihood to face the scenario (new feature, I think landed in 2.4 or 2.5). I didn't follow it, so don't know the requirement behind it
- This is supposed to be a short-term fix in our code right? if we are confident a long term solution can arrive in the next or 2 releases, we might be ok taking this risk

Probably it's worth to include PMs as well in the decision.

Reply:
This PR might be revisit in the future if we can remove the NP entirely so let's focus on "DS project" for now.
For "advanced" use cases where the user creates a namespace explicitly we can document the necessary NP

Reply:
sounds good, do we already know what would be the NP in case of "advanced" use case? So we can provide it in the docs.

Reply:
Asked Adam to confirm https://issues.redhat.com/browse/RHOAIENG-1955?focusedId=23833955&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-23833955 but I would keep "DS project" filter in this PR

Reply:
IMHO the most important use-case is the one from @cfchase. If I understood it correctly: assuming a namespace belongs to a tenant, services within the namespace should be reachable between them transparently, regardless if they belong or not to the mesh.

Also, we should have NPs to let infrastructure components to do their job. This would mean to allow pods in the `AppNamespace` to access pods in the namespaces where a KServe `InferenceService` exists (regardless if it is a DSProject or not). I think this PR already covers this, right?.

To me, traffic flowing to/from other DSProject is just a "plus" here. Assuming that the discussion here is about cross-namespace traffic between DSProjects and non-DSProjects, I'd agree that for the time being documenting it as a known issue with the workaround would be OK. However, this PR should still properly enable traffic between services in the same namespace where an `InferenceService` exists, even if the namespace in play is not a DSProject.